### PR TITLE
Add basic region, block, operation iteration

### PR DESCRIPTION
**Pull request by jpienaar**
Start of a basic walk from operation to region to block to operation.

### CODE REVIEW

**Comment by jpienaar** [comment]
I was also considering just exposing a list return here, that may be more convenient Native side, could switch to that instead

File: test/MLIR/NativeSpec.hs [hunk 1]
```diff
@@ -100,6 +103,24 @@ spec = do
...
+      MLIR.destroyModule exampleModule
+
+    it "Can show operations inside region of function" $ \ctx -> do
+      exampleModule <- liftM fromJust $
+        MLIR.withStringRef exampleModuleStr $ MLIR.parseModule ctx
+      operations <- (MLIR.getModuleBody >=> MLIR.getBlockOperations) exampleModule
+      blocks <- ((MLIR.getOperationRegion 0) >=> MLIR.getRegionBlocks) (head operations)
+      ops <- MLIR.getBlockOperations $ head blocks
+      opStrs <- sequence $ map MLIR.showOperation ops
+      (BS.intercalate " ; " opStrs) `shouldBe` (pack "%0 = arith.addi %arg0, %arg0 : i32 ; return %0 : i32")
```
Comment (apaszke) [inline test/MLIR/NativeSpec.hs:121]: Do you need this `pack`? I would assume that with overloaded string literals it should work out fine.

File: src/MLIR/Native.hs [hunk 2]
```diff
@@ -303,6 +348,16 @@ showSomething action = do
       free dataPtr
       return bs
 
+-- | Unroll using a function that is equivalent to "get next" inside IO.
+unrollIOMaybe :: (a -> IO (Maybe a)) -> IO (Maybe a) -> IO [a]
+unrollIOMaybe fn z = do
+  x <- z
+  case x of
+      Nothing -> return mzero
```
Comment (apaszke) [inline src/MLIR/Native.hs:356]: I think that this `mzero` can be just `[]`

File: src/MLIR/Native.hs [hunk 3]
```diff
@@ -303,6 +348,16 @@ showSomething action = do
...
+      Just x' -> do
+          xs <- unrollIOMaybe fn (fn x')
+          return (return x' `mplus` xs)
```
Comment (apaszke) [inline src/MLIR/Native.hs:359]: Perhaps just `return $ x' : xs`?

Comment (apaszke) [reply src/MLIR/Native.hs:359]: And if you want to be very fancy then you can fold this whole `do` block into `(x':) <$> unrollIOMaybe fn (fn x')`

**Review by apaszke — APPROVED** [review]
Nice!

**Pull request merged**